pilot/pkg/config/kube/gateway/conversion.go

```diff
@@ -410,23 +410,25 @@ func buildHTTPVirtualServices(
 			if len(matches) == 0 {
 				matches = append(matches, nil)
 			}
+			r.Matches = make([]k8s.HTTPRouteMatch, 0, len(matches))
 			for _, m := range matches {
 				if m != nil {
-					r.Matches = []k8s.HTTPRouteMatch{*m}
-				}
-				vs, err := convertHTTPRoute(r, ctx, obj, n, !mesh)
-				// This was a hard error
-				if vs == nil {
-					res.error = err
-					return conversionResult{error: err}
-				}
-				// Got an error but also routes
-				if err != nil {
-					res.error = err
+					r.Matches = append(r.Matches, *m)
```

Comment:
Comment: 'split the rule to make sure each rule has up to one match'

This seems inconsitent?

Reply:
Which comment are you referring to?

Reply:
I'm curious why we need to get the reference in L408 and dereference it in L415?

Reply:
Does grpc have the same problem? 

https://github.com/istio/istio/blob/0edb8f572207fcef2797d403b141539fcb9a7d0e/pilot/pkg/config/kube/gateway/conversion.go#L641

Reply:
Very likely, will take a look later.

Reply:
Isn't the structure of VS with new logic looks a lot more like the original HTTPRoute structure?

Reply:
the goal is not to make the structures look similar but to make things behave semantically correct. HTTPRoute specifies a sorting order. This is per-match, not per-rule. Istio evaluates rules in order. 

We need to split them out so we can correctly sort on a per match basis rather than a per-rule basis

Reply:
fine, let me close this first.